File: spikes/builtin-host-boundary/s1-orphan/probe.mjs
// S1 probe (node). naive = what js/src/builtin.ts does today (spawn shell:true + child.kill).
// group = detached:true + process.kill(-pid).
import { spawn } from "node:child_process"
const [mode, marker] = process.argv.slice(2)
const command = `sleep 0.2; sh -c 'sleep 1; touch ${marker}'`
const child = spawn(command, { shell: true, detached: mode === "group" })
setTimeout(() => {
  if (mode === "naive") child.kill("SIGKILL")
  else {
    try { process.kill(-child.pid, "SIGTERM") } catch {}
    setTimeout(() => { try { process.kill(-child.pid, "SIGKILL") } catch {} }, 200)
  }
}, 300)
child.on("close", () => { console.log("killed"); process.exit(0) })

File: spikes/builtin-host-boundary/s1-orphan/probe.py
# S1 probe (python). naive = what python/src/toolnexus/builtin.py does today
# (subprocess.run(shell=True, timeout=...)). group = start_new_session + killpg.
import os, signal, subprocess, sys, time
mode, marker = sys.argv[1], sys.argv[2]
command = f"sleep 0.2; sh -c 'sleep 1; touch {marker}'"
if mode == "naive":
    try:
        subprocess.run(command, shell=True, timeout=0.3, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    except subprocess.TimeoutExpired:
        pass
else:
    p = subprocess.Popen(command, shell=True, start_new_session=True,
                         stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    try:
        p.communicate(timeout=0.3)
    except subprocess.TimeoutExpired:
        os.killpg(p.pid, signal.SIGTERM)
        time.sleep(0.2)
        try:
            os.killpg(p.pid, signal.SIGKILL)
        except ProcessLookupError:
            pass
        p.wait()
print("killed")
